gtp2ogs.js

```diff
@@ -1017,10 +1021,36 @@ class Game {
         if (this.state.phase != 'play')
             return;
 
-	let sendMove = (moves) => {  this.uploadMove(moves[0]);  };
-	let sendPass = ()      => {  this.uploadMove({'x': -1});  };
+	let sendPass = () => {  this.uploadMove({'x': -1});  };
+	let doing_handicap = (this.state.free_handicap_placement && this.state.handicap > 1 &&
+			      this.state.moves.length < this.state.handicap);
 	
-	this.getBotMoves("genmove " + this.my_color, sendMove, sendPass);
+	if (!doing_handicap) {  // Regular genmove ...
+	    let sendMove = (moves) => {  this.uploadMove(moves[0]);  };
+	    this.getBotMoves("genmove " + this.my_color, sendMove, sendPass);
+	    return;
+	}
+	
+	// Already have handicap stones ? Return next one.
+	if (this.handicap_moves.length) {
+	    this.uploadMove(this.handicap_moves.shift());
+	    return;
+	}
+
+	// Get handicap stones from bot and return first one.
+	let storeMoves = (moves) => {
+	    if (moves.length != this.state.handicap) {  // Sanity check
+		this.log("place_free_handicap returned wrong number of handicap stones, resigning.");
+		if (this.bot) this.bot.kill();
+		this.bot = null;
+		this.uploadMove({'resign': true});
+		return;
+	    }
+	    this.handicap_moves = moves;
+	    this.uploadMove(this.handicap_moves.shift());
+	};
+	    
+	this.getBotMoves("place_free_handicap " + this.state.handicap, storeMoves, sendPass);
```

Comment:
You should probably add a command line flag to request the use of `place_free_handicap`, I think @wonderingabout was worried that phoenixgo might not implement that GTP command?

Reply:
Then phoenixgo is broken. Please don't use gtp2ogs to work around bugs in bots ! If a bot admin wants to add such code to his copy of gtp2ogs he is free to do so, but i wouldn't want that in the repository. At any rate gtp2ogs should implement the gtp spec correctly by default (use place_free_handicap).

Reply:
phoenixgo supports free handicap as is currently on devel branch, as black and as white without any issue (to give or take handi)

the only thing i pointed is the behaviour of human waiting long or leaving games waiting for bot to play while they didnt put all handicap stones, as explained in #80 

you are right that we shouldnt hack gtp2ogs, custom branches can achieve that